modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,701 @@
+//
+
+/* Return a vector (1D array) over domain ``{0..#length}``*/
+proc Vector(length, type eltType=real) {
+  //if (length <= 0) then compilerError("Vector length must be > 0");
+  var V: [0..#length] eltType;
+  return V;
+}
+
+
+/* Return a vector (1D array) over domain ``Dom`` */
+proc Vector(Dom: domain(1), type eltType=real) {
+  var V: [Dom] eltType;
+  return V;
+}
+
+
+/* Return a vector (1D array) with domain and values of ``A`` */
+proc Vector(A: [?Dom] ?Atype, type eltType=Atype ) {
+  var V: [Dom] eltType = A;
+  return V;
+}
+
+
+pragma "no doc"
+proc Vector(x: ?t, Scalars...?n)  where isNumericType(t) {
+  type eltType = Scalars(1).type;
+  return Vector(x, (...Scalars), eltType=eltType);
+}
+
+
+/* Return a vector (1D array), given 2 or more numeric values
+
+   If `type` is omitted, it will be inferred from the first argument
+*/
+proc Vector(x: ?t, Scalars...?n, type eltType) where isNumericType(t) {
+
+  if !isNumeric(Scalars(1)) then
+      compilerError("Vector() expected numeric arguments");
+
+  var V: [0..n] eltType;
+
+  V[0] = x: eltType;
+
+  for i in 1..n do V[i] = Scalars[i]: eltType;
+
+  return V;
+}
+
+
+/* Return a square matrix (2D array) over domain ``{0..#rows, 0..#rows}``*/
+proc Matrix(rows, type eltType=real) where isIntegral(rows) {
+  if rows <= 0 then halt("Matrix dimensions must be > 0");
+  var M: [0..#rows, 0..#rows] eltType;
+  return M;
+}
+
+
+/* Return a matrix (2D array) over domain ``{0..#rows, 0..#cols}``*/
+proc Matrix(rows, cols, type eltType=real) where isIntegral(rows) && isIntegral(cols) {
+  if rows <= 0 || cols <= 0 then halt("Matrix dimensions must be > 0");
+  var M: [0..#rows, 0..#cols] eltType;
+  return M;
+}
+
+
+/* Return a matrix (2D array) over domain ``Dom`` */
+proc Matrix(Dom: domain(2), type eltType=real) where Dom.rank == 2 {
+  var M: [Dom] eltType;
+  return M;
+}
+
+
+/* Return a matrix (2D array) with domain and values of ``A`` */
+proc Matrix(A: [?Dom] ?eltType) where Dom.rank == 2 {
+  var M: [Dom] eltType = A;
+  return M;
+}
+
+
+pragma "no doc"
+proc Matrix(Arrays...?n) {
+  type eltType = Arrays(1).eltType;
+  return Matrix((...Arrays), eltType=eltType);
+}
+
+
+/*
+    Return a matrix (2D array), given 2 or more vectors. Vectors are
+    concatenated such that the ``ith`` vector corresponds to the matrix slice:
+    ``A[i, ..]``
+
+    If `type` is omitted, it will be inferred from the first argument
+*/
+proc Matrix(const Arrays...?n, type eltType) {
+  // TODO -- assert all array domains are same length
+  //         Can this be done via type query?
+
+  if Arrays(1).domain.rank != 1 then compilerError("Matrix() expected 1D arrays");
+
+  const dim2 = 0..#Arrays(1).domain.dim(1).length,
+        dim1 = 0..#n;
+
+  var M: [{dim1, dim2}] eltType;
+
+  for i in dim1 do
+    M[i, ..] = Arrays(i+1)[..]: eltType;
+
+  return M;
+}
+
+
+/* Return an identity matrix over domain ``{0..#m, 0..#n}`` */
+proc eye(m, n, type eltType=real) {
```

Comment:
This library was primarily based on the linear algebra support in numpy and matlab. I'm open to alternatives to `eye`, but let's leave it as is for now, and discuss it further before code/doc.